### ISSUE TITLE
Multi-GPU support

### ISSUE THREAD

**Issue by soumith** [torch/cutorch#42]
MultiGPU support has been implemented in cutorch (and by extension all torch cuda libraries like cunn, cudnn etc.).
- Switch the device on the fly with cutorch.setDevice(devID)
- All cuda calls are asynchronous, and can be synchronized with cutorch.synchronize()

Example usage for tensors:

``` lua
-- Let us do matrix addition for matrices sitting on two different GPUs
cutorch.setDevice(1)
matrix1 = torch.CudaTensor(10):fill(1)
print(matrix1) -- printing is a synchronous call, so you dont have to explicitly call cutorch.synchronize()
cutorch.setDevice(2)
matrix2 = torch.CudaTensor(10):fill(2)
print(matrix2) 
matrix2:add(matrix1) -- matrix1 is seamlessly copied onto GPU2 and added to matrix2
print(matrix2)
```

if you want to do data-parallel training of neural nets (including convnets), your training loop can run like this:

For each mini-batch: 

``` lua
1. load data (preferably using multiple threads, for example using [threads-ffi](https://github.com/torch/threads-ffi))
2. loop over GPUs (the loop below will be completely anynchronous, so will run parallely)
  2.1. model[gpuX]:forward
  2.2. criterion[gpuX]:forward
  2.3. criterion[gpuX]:backward
  2.4. model[gpuX]:backward
3. cutorch.synchronize()
4. accumulate GPUx's gradParameters to GPU1's gradParameters
5. do SGD on GPU1
6. copy back GPU1's parameters to GPUx
7. cutorch.synchronize() and print accuracy etc.
```

Loop back to 1 for next mini-batch

Also, to train ConvNets using multiple GPUs, I recommend using [CuDNN](https://github.com/soumith/cudnn.torch) for the convolution layers, as I've tested that they are completely asynchronous (meaning that the processing runs parallely on multiple GPUs)

**Comments below describe the technical details of changes made. If you just want to use Multi-GPU, you can stop reading now.**

**Comment by algred** [comment]
Error happens when I run the  example: it seems the math operations are not "seamless" as declared:

th> cutorch.setDevice(1)
th> matrix1 = torch.CudaTensor(10):fill(1)
th> print(matrix1) 
cutorch.synchronize()
 1
 1
 1
 1
 1
 1
 1
 1
 1
 1
[torch.CudaTensor of size 10]
th> cutorch.setDevice(2)
th> matrix2 = torch.CudaTensor(10):fill(2)
th> print(matrix2) 
 2
 2
 2
 2
 2
 2
 2
 2
 2
 2
[torch.CudaTensor of size 10]
th> matrix2:add(matrix1) 
[string "matrix2:add(matrix1) -- matrix1 is seamlessly..."]:1: Assertion `THCudaTensor_checkGPU(state, 3, self_, src1, src2)' failed.  at /tmp/luarocks_cutorch-scm-1-6658/cutorch/lib/THC/THCTensorMathPointwise.cu:82
stack traceback:
        [C]: in function 'add'
        [string "matrix2:add(matrix1) -- matrix1 is seamlessly..."]:1: in main chunk
        [C]: in function 'xpcall'
        /home/shugao/torch/install/share/lua/5.1/trepl/init.lua:648: in function 'repl'
        ...ugao/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
        [C]: at 0x00406670

**Comment by soumith** [comment]
@algred the only operation that is allowed cross-GPU is the copy operation. All other operations are checked with assertions to make sure what you are doing is not possible. To get good performance, you'll have to copy matrix1 onto matrix2's GPU and then do the mathematical operation.

If you dont like this setting, you can simply disable these assertions by adding the define DISABLE_CHECK_GPU and reinstalling cutorch.

https://github.com/torch/cutorch/blob/master/lib/THC/THCTensor.c#L761

**Comment by algred** [comment]
@soumith Thank you very much for replying! 

So this means to enable the Multi-GPU support as described above, the first thing need to do is DISABLE_CHECK_GPU and reinstalling cutorch? I am quite new to Torch, and can you give me some hints on how to reinstall after I change the code (i.e. adding the define)?

**Comment by darksigma** [comment]
Does this PR support GPU Direct RDMA? Or are additional lower-level modifications necessary to run on a multi GPU Mellanox/GTX Titan X cluster?

**Comment by eriche2016** [comment]
If i do data parallel training, how does the minibatch data are split over the multi GPUs, do they split evenly by the scheduler, or just split the minibatch manually

**Comment by soumith** [comment]
evenly

**Comment by soumith** [comment]
@darksigma try looking at nccl.torch for that.

**Comment by byronwwang** [comment]
@soumith For mini-batch splitting, is there any shuffle before it? Or just split the batch evenly according to the original order of samples in the batch.

**Comment by soumith** [comment]
No shuffle, split evenly
